File: projectfund-main/smart-contracts/scripts/add-authority.js
const hre = require("hardhat");
require("dotenv").config();

async function main() {
  console.log("\n🔐 Adding Authority to PublicFundManagement Contract...\n");
  
  // Get the deployer/owner wallet
  const [deployer] = await hre.ethers.getSigners();
  console.log("👤 Owner account:", deployer.address);
  
  // Contract address
  const contractAddress = "0x9B00068CfBF060E4aad61a892a86E98C108D760e";
  console.log("📍 Contract address:", contractAddress);
  
  // The address you want to add as authority (your MetaMask wallet)
  const authorityAddress = "0x6459c8db925694d0b376980239ff00a2eeeba311"; // New authority address
  console.log("➕ Adding authority:", authorityAddress);
  
  // Get contract instance
  const contractABI = [
    "function addAuthority(address authority) external",
    "function authorities(address) external view returns (bool)",
    "function admin() external view returns (address)"
  ];
  const contract = new hre.ethers.Contract(contractAddress, contractABI, deployer);
  
  try {
    // Check if already an authority
    console.log("\n🔍 Checking current authority status...");
    const isAuthority = await contract.authorities(authorityAddress);
    if (isAuthority) {
      console.log("✅ Address is already an authority!\n");
      return;
    }
    
    // Check admin
    const admin = await contract.admin();
    console.log("👑 Contract admin:", admin);
    
    if (admin.toLowerCase() !== deployer.address.toLowerCase()) {
      console.log("\n❌ ERROR: You are not the contract admin!");
      console.log("Only the admin can add authorities.");
      console.log("Current deployer:", deployer.address);
      console.log("Contract admin:", admin);
      console.log("\n💡 TIP: Make sure your .env file has the correct PRIVATE_KEY");
      console.log("The private key should be from the wallet that deployed the contract.");
      process.exitCode = 1;
      return;
    }
    
    // Add authority
    console.log("\n⏳ Adding authority...");
    const tx = await contract.addAuthority(authorityAddress);
    
    console.log("📝 Transaction hash:", tx.hash);
    console.log("🔗 View on Etherscan: https://sepolia.etherscan.io/tx/" + tx.hash);
    
    // Wait for confirmation
    console.log("\n⏳ Waiting for confirmation...");
    const receipt = await tx.wait();
    
    if (receipt.status === 1) {
      console.log("✅ Transaction CONFIRMED!\n");
      
      // Verify
      const isAuthorityNow = await contract.authorities(authorityAddress);
      if (isAuthorityNow) {
        console.log("🎉 SUCCESS! Address is now an authority!");
        console.log("✅", authorityAddress, "can now:");
        console.log("   • Create proposals");
        console.log("   • Approve SBT applications");
        console.log("   • Release funds\n");
      } else {
        console.log("⚠️ Warning: Could not verify authority status");
      }
    } else {
      console.log("❌ Transaction FAILED");
      process.exitCode = 1;
    }
  } catch (error) {
    console.error("\n❌ Error:", error.message);
    if (error.message.includes("admin") || error.message.includes("owner")) {
      console.log("\n💡 Make sure you're using the deployer's private key in your .env file");
    }
    process.exitCode = 1;
  }
}

main().catch((error) => {
  console.error("\n❌ Error:", error.message);
  process.exitCode = 1;
});

File: projectfund-main/backend/add_authority.py
#!/usr/bin/env python3
"""
Add authority to PublicFundManagement contract
"""
import os
import sys
from web3 import Web3

# Configuration
PROVIDER_URL = "http://127.0.0.1:8545"
CONTRACT_ADDRESS = "0x9B00068CfBF060E4aad61a892a86E98C108D760e"
AUTHORITY_TO_ADD = "0x6459c8db925694d0b376980239ff00a2eeeba311"
ADMIN_PRIVATE_KEY = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"

# Contract ABI
CONTRACT_ABI = [
    {
        "inputs": [{"internalType": "address", "name": "authority", "type": "address"}],
        "name": "addAuthority",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "address", "name": "", "type": "address"}],
        "name": "authorities",
        "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
        "stateMutability": "view",
        "type": "function"
    }
]

def main():
    print("\n🔐 Adding Authority to PublicFundManagement Contract...\n")
    
    # Connect to provider
    w3 = Web3(Web3.HTTPProvider(PROVIDER_URL))
    
    if not w3.is_connected():
        print("❌ Error: Cannot connect to blockchain node at", PROVIDER_URL)
        print("Make sure Hardhat node is running: npx hardhat node")
        sys.exit(1)
    
    print("✅ Connected to blockchain node")
    
    # Setup account
    admin_account = w3.eth.account.from_key(ADMIN_PRIVATE_KEY)
    print(f"👤 Admin account: {admin_account.address}")
    print(f"📍 Contract: {CONTRACT_ADDRESS}")
    print(f"➕ Adding as authority: {AUTHORITY_TO_ADD}")
    
    # Get contract instance
    contract = w3.eth.contract(address=CONTRACT_ADDRESS, abi=CONTRACT_ABI)
    
    try:
        # Check if already an authority
        print("\n🔍 Checking current status...")
        is_authority = contract.functions.authorities(AUTHORITY_TO_ADD).call()
        
        if is_authority:
            print("✅ Already an authority!\n")
            return
        
        print("⏳ Not yet an authority, adding...\n")
        
        # Build transaction
        tx_func = contract.functions.addAuthority(AUTHORITY_TO_ADD)
        
        # Estimate gas
        gas_estimate = tx_func.estimate_gas({"from": admin_account.address})
        print(f"📊 Estimated gas: {gas_estimate}")
        
        # Get nonce
        nonce = w3.eth.get_transaction_count(admin_account.address)
        
        # Get gas price
        gas_price = w3.eth.gas_price
        print(f"💰 Gas price: {gas_price} wei")
        
        # Build transaction
        tx_dict = tx_func.build_transaction({
            "from": admin_account.address,
            "nonce": nonce,
            "gas": gas_estimate,
            "gasPrice": gas_price,
        })
        
        # Sign transaction
        signed_tx = w3.eth.account.sign_transaction(tx_dict, ADMIN_PRIVATE_KEY)
        
        # Send transaction
        print("\n📤 Sending transaction...")
        tx_hash = w3.eth.send_raw_transaction(signed_tx.rawTransaction)
        print(f"📝 Transaction hash: {tx_hash.hex()}")
        
        # Wait for confirmation
        print("⏳ Waiting for confirmation...")
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=30)
        
        if receipt['status'] == 1:
            print("✅ Transaction CONFIRMED!\n")
            
            # Verify
            is_authority_now = contract.functions.authorities(AUTHORITY_TO_ADD).call()
            if is_authority_now:
                print("🎉 SUCCESS! Address is now an authority!")
                print(f"✅ {AUTHORITY_TO_ADD} can now:\n")
                print("   • Create proposals")
                print("   • Approve SBT applications")
                print("   • Release funds\n")
            else:
                print("⚠️  Warning: Could not verify authority status")
        else:
            print("❌ Transaction FAILED")
            sys.exit(1)
            
    except Exception as e:
        print(f"❌ Error: {str(e)}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    main()
